File: tools/schema/validate.py
"""Lightweight JSON config validator using schema definitions from schemas.py.

No external dependencies — pure Python stdlib.
"""

import json
from pathlib import Path
from typing import Any

from tools.schema.schemas import PROVE_SCHEMA, SETTINGS_SCHEMA

TYPE_MAP = {
    "str": str,
    "int": (int, float),
    "bool": bool,
    "list": list,
    "dict": dict,
}


class ValidationError:
    """A single validation finding."""

    def __init__(self, path: str, message: str, severity: str = "error"):
        self.path = path
        self.message = message
        self.severity = severity  # "error" or "warning"

    def __str__(self) -> str:
        prefix = "ERROR" if self.severity == "error" else "WARN"
        return f"  {prefix}: {self.path}: {self.message}"

    def __repr__(self) -> str:
        return f"ValidationError({self.path!r}, {self.message!r})"


def _validate_value(value: Any, spec: dict, path: str) -> list[ValidationError]:
    """Validate a single value against a field spec."""
    errors: list[ValidationError] = []
    expected_type = spec.get("type", "any")

    if expected_type == "any":
        return errors

    # Type check
    python_type = TYPE_MAP.get(expected_type)
    if python_type and not isinstance(value, python_type):
        # Allow int for float
        if expected_type == "int" and isinstance(value, float) and value == int(value):
            pass
        else:
            errors.append(
                ValidationError(
                    path,
                    f"expected {expected_type}, got {type(value).__name__}",
                )
            )
            return errors  # Skip further checks if type is wrong

    # Enum check
    if "enum" in spec and value not in spec["enum"]:
        errors.append(
            ValidationError(
                path,
                f"must be one of {spec['enum']}, got {value!r}",
            )
        )

    # List items
    if expected_type == "list" and "items" in spec:
        for i, item in enumerate(value):
            errors.extend(_validate_value(item, spec["items"], f"{path}[{i}]"))

    # Dict fields (known keys)
    if expected_type == "dict" and "fields" in spec:
        errors.extend(_validate_fields(value, spec["fields"], path))

    # Dict values (arbitrary keys, uniform value type)
    if expected_type == "dict" and "values" in spec and "fields" not in spec:
        for key, val in value.items():
            errors.extend(_validate_value(val, spec["values"], f"{path}.{key}"))

    return errors


def _validate_fields(
    data: dict, fields: dict, prefix: str = ""
) -> list[ValidationError]:
    """Validate a dict's fields against schema field definitions."""
    errors: list[ValidationError] = []

    for field_name, spec in fields.items():
        path = f"{prefix}.{field_name}" if prefix else field_name

        if field_name not in data:
            if spec.get("required", False):
                errors.append(ValidationError(path, "required field is missing"))
            continue

        errors.extend(_validate_value(data[field_name], spec, path))

    # Warn about unknown top-level fields (not an error — extensibility)
    known = set(fields.keys())
    for key in data:
        if key not in known:
            path = f"{prefix}.{key}" if prefix else key
            errors.append(
                ValidationError(
                    path,
                    "unknown field (not in schema — may be from a tool or future version)",
                    severity="warning",
                )
            )

    return errors


def validate_config(
    config: dict, schema: dict, strict: bool = False
) -> list[ValidationError]:
    """Validate a config dict against a schema.

    Args:
        config: The parsed JSON config
        schema: Schema definition (PROVE_SCHEMA or SETTINGS_SCHEMA)
        strict: If True, treat warnings as errors

    Returns:
        List of ValidationError objects
    """
    errors = _validate_fields(config, schema["fields"])

    if strict:
        for e in errors:
            if e.severity == "warning":
                e.severity = "error"

    return errors


def validate_file(
    path: str, schema: dict | None = None, strict: bool = False
) -> tuple[dict | None, list[ValidationError]]:
    """Validate a JSON config file.

    Auto-detects schema from filename if not provided.

    Returns:
        (parsed_config_or_None, errors)
    """
    filepath = Path(path)

    if not filepath.exists():
        return None, [ValidationError(path, "file not found")]

    try:
        with open(filepath) as f:
            config = json.load(f)
    except json.JSONDecodeError as e:
        return None, [ValidationError(path, f"invalid JSON: {e}")]

    if not isinstance(config, dict):
        return None, [ValidationError(path, "top-level value must be an object")]

    # Auto-detect schema
    if schema is None:
        if filepath.name == ".prove.json" or str(filepath).endswith(".claude/.prove.json"):
            schema = PROVE_SCHEMA
        elif filepath.name == "settings.json" and ".claude" in str(filepath):
            schema = SETTINGS_SCHEMA
        else:
            return config, [
                ValidationError(
                    path,
                    "cannot auto-detect schema — pass schema explicitly",
                    severity="warning",
                )
            ]

    errors = validate_config(config, schema, strict=strict)
    return config, errors

File: packages/cli/src/topics/schema/validate.ts
/**
 * Lightweight JSON config validator. Pure stdlib — `node:fs` + `node:path`.
 *
 * Ported 1:1 from `tools/schema/validate.py`. Behavior parity:
 *   - Type mismatches produce an error and abort further checks on that node.
 *   - Enum violations produce an error (only when type check passes).
 *   - Required fields absent from the data produce an error.
 *   - Unknown keys on a dict-with-`fields` produce a WARNING (unchanged by
 *     normal runs; `strict: true` promotes warnings to errors in place).
 *   - Unknown-field warnings are emitted AFTER required-field errors for
 *     stable ordering in test output.
 *
 * Path construction mirrors the Python version: `prefix.field` for dicts,
 * `prefix[index]` for list items. `ValidationError.toString()` is byte-
 * identical to the Python `__str__`.
 */

import { readFileSync } from 'node:fs';
import { basename } from 'node:path';
import { type FieldSpec, PROVE_SCHEMA, SETTINGS_SCHEMA, type Schema } from './schemas';

export type Severity = 'error' | 'warning';

/** A single validation finding. */
export class ValidationError {
  constructor(
    public path: string,
    public message: string,
    public severity: Severity = 'error',
  ) {}

  toString(): string {
    const prefix = this.severity === 'error' ? 'ERROR' : 'WARN';
    return `  ${prefix}: ${this.path}: ${this.message}`;
  }
}

/**
 * Validate a config object against a schema.
 *
 * @param config Parsed JSON object (must already be a plain object).
 * @param schema `PROVE_SCHEMA` or `SETTINGS_SCHEMA`.
 * @param strict When true, every warning is promoted to an error in place.
 */
export function validateConfig(
  config: Record<string, unknown>,
  schema: Schema,
  strict = false,
): ValidationError[] {
  const errors = validateFields(config, schema.fields, '');
  if (strict) {
    for (const e of errors) {
      if (e.severity === 'warning') {
        e.severity = 'error';
      }
    }
  }
  return errors;
}

export interface ValidateFileResult {
  config: Record<string, unknown> | null;
  errors: ValidationError[];
}

/**
 * Read and validate a JSON config file on disk.
 *
 * Auto-detects the schema from the path when `schema` is omitted:
 *   - basename `.prove.json` OR path ending `.claude/.prove.json` -> PROVE_SCHEMA
 *   - basename `settings.json` with `.claude` in the path        -> SETTINGS_SCHEMA
 *   - anything else -> a single warning, config returned unvalidated.
 *
 * Emits an error (with `config: null`) for missing file, invalid JSON, or a
 * non-object top-level value.
 */
export function validateFile(path: string, schema?: Schema, strict = false): ValidateFileResult {
  let raw: string;
  try {
    raw = readFileSync(path, 'utf8');
  } catch {
    return { config: null, errors: [new ValidationError(path, 'file not found')] };
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return {
      config: null,
      errors: [new ValidationError(path, `invalid JSON: ${msg}`)],
    };
  }

  if (!isPlainObject(parsed)) {
    return {
      config: null,
      errors: [new ValidationError(path, 'top-level value must be an object')],
    };
  }

  const config = parsed;
  const resolved = schema ?? autoDetectSchema(path);
  if (resolved === null) {
    return {
      config,
      errors: [
        new ValidationError(path, 'cannot auto-detect schema — pass schema explicitly', 'warning'),
      ],
    };
  }

  return { config, errors: validateConfig(config, resolved, strict) };
}

// --- internals ---

function autoDetectSchema(path: string): Schema | null {
  const name = basename(path);
  if (name === '.prove.json' || path.endsWith('.claude/.prove.json')) {
    return PROVE_SCHEMA;
  }
  if (name === 'settings.json' && path.includes('.claude')) {
    return SETTINGS_SCHEMA;
  }
  return null;
}

function validateFields(
  data: Record<string, unknown>,
  fields: Record<string, FieldSpec>,
  prefix: string,
): ValidationError[] {
  const errors: ValidationError[] = [];
  const known = new Set(Object.keys(fields));

  for (const [fieldName, spec] of Object.entries(fields)) {
    const path = prefix ? `${prefix}.${fieldName}` : fieldName;
    if (!(fieldName in data)) {
      if (spec.required) {
        errors.push(new ValidationError(path, 'required field is missing'));
      }
      continue;
    }
    errors.push(...validateValue(data[fieldName], spec, path));
  }

  // Unknown-field warnings are emitted AFTER required-field errors so the
  // ordering of test output stays stable (matches Python source).
  for (const key of Object.keys(data)) {
    if (known.has(key)) continue;
    const path = prefix ? `${prefix}.${key}` : key;
    errors.push(
      new ValidationError(
        path,
        'unknown field (not in schema — may be from a tool or future version)',
        'warning',
      ),
    );
  }

  return errors;
}

function validateValue(value: unknown, spec: FieldSpec, path: string): ValidationError[] {
  const errors: ValidationError[] = [];
  const expected = spec.type;

  if (expected === 'any') return errors;

  if (!matchesType(value, expected)) {
    errors.push(new ValidationError(path, `expected ${expected}, got ${typeName(value)}`));
    return errors; // Skip further checks if type is wrong.
  }

  if (spec.enum && !spec.enum.includes(value as string | number | boolean)) {
    errors.push(
      new ValidationError(
        path,
        `must be one of ${formatEnum(spec.enum)}, got ${formatValue(value)}`,
      ),
    );
  }

  if (expected === 'list' && spec.items) {
    const arr = value as unknown[];
    for (let i = 0; i < arr.length; i++) {
      errors.push(...validateValue(arr[i], spec.items, `${path}[${i}]`));
    }
  }

  if (expected === 'dict' && spec.fields) {
    errors.push(...validateFields(value as Record<string, unknown>, spec.fields, path));
  }

  if (expected === 'dict' && spec.values && !spec.fields) {
    for (const [key, val] of Object.entries(value as Record<string, unknown>)) {
      errors.push(...validateValue(val, spec.values, `${path}.${key}`));
    }
  }

  return errors;
}

/**
 * Type predicate mirroring Python's TYPE_MAP semantics. Notable quirks kept
 * intentional:
 *   - `int` accepts any JS number that is an integer (includes integer-valued
 *     floats, matching Python's `int(value) == value` check).
 *   - `bool` and `int` are disjoint here: a boolean NEVER satisfies `int`.
 *     (Python conflates the two because `bool` subclasses `int`, but the
 *     test_validate.py suite relies on the separation — e.g. `enabled: "yes"`
 *     is rejected against `type: bool` without ambiguity.)
 *   - `list` = plain array; `dict` = plain object (not array, not null).
 */
function matchesType(value: unknown, expected: Exclude<FieldType, 'any'>): boolean {
  switch (expected) {
    case 'str':
      return typeof value === 'string';
    case 'bool':
      return typeof value === 'boolean';
    case 'int':
      return typeof value === 'number' && Number.isInteger(value);
    case 'list':
      return Array.isArray(value);
    case 'dict':
      return isPlainObject(value);
  }
}

type FieldType = FieldSpec['type'];

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/**
 * Render a value the way Python's `repr()` would for the message-building
 * sites in validate.py: strings get single quotes, everything else falls
 * back to JSON.
 */
function formatValue(value: unknown): string {
  if (typeof value === 'string') return `'${value}'`;
  return JSON.stringify(value);
}

/**
 * Render a list literal the way Python's `str(list)` would — single-quoted
 * strings inside brackets — so error messages match the Python source
 * byte-for-byte: `["build", "lint"]` -> `['build', 'lint']`.
 */
function formatEnum(values: readonly (string | number | boolean)[]): string {
  const parts = values.map((v) => formatValue(v));
  return `[${parts.join(', ')}]`;
}

/**
 * Mirror Python's `type(value).__name__` for the common JSON types. Used
 * only in type-mismatch messages.
 */
function typeName(value: unknown): string {
  if (value === null) return 'NoneType';
  if (Array.isArray(value)) return 'list';
  const t = typeof value;
  switch (t) {
    case 'string':
      return 'str';
    case 'boolean':
      return 'bool';
    case 'number':
      return Number.isInteger(value) ? 'int' : 'float';
    case 'object':
      return 'dict';
    default:
      return t;
  }
}
